src/Microsoft.VisualStudio.Mac.LanguageServices.Razor/ProjectSystem/RazorProjectHostBase.cs

```diff
@@ -0,0 +1,166 @@
+        private const string ProjectChangedHint = "References";
+
+        private bool _batchingProjectChanges;
+        private readonly DotNetProject _dotNetProject;
+        private readonly ForegroundDispatcher _foregroundDispatcher;
+        private readonly ProjectSnapshotManagerBase _projectSnapshotManager;
+        private readonly SemaphoreSlim _projectChangedSemaphore;
+        private HostProject _currentHostProject;
+
+        public RazorProjectHostBase(
+            DotNetProject project,
+            ForegroundDispatcher foregroundDispatcher,
+            ProjectSnapshotManagerBase projectSnapshotManager)
+        {
+            if (project == null)
+            {
+                throw new ArgumentNullException(nameof(project));
+            }
+
+            if (foregroundDispatcher == null)
+            {
+                throw new ArgumentNullException(nameof(foregroundDispatcher));
+            }
+
+            if (projectSnapshotManager == null)
+            {
+                throw new ArgumentNullException(nameof(projectSnapshotManager));
+            }
+
+            _dotNetProject = project;
+            _foregroundDispatcher = foregroundDispatcher;
+            _projectSnapshotManager = projectSnapshotManager;
+            _projectChangedSemaphore = new SemaphoreSlim(initialCount: 1);
+
+            AttachToProject();
+        }
+
+        public DotNetProject DotNetProject => _dotNetProject;
+
+        public HostProject HostProject => _currentHostProject;
+
+        protected ForegroundDispatcher ForegroundDispatcher => _foregroundDispatcher;
+
+        public void Detatch()
+        {
+            _foregroundDispatcher.AssertForegroundThread();
+
+            DotNetProject.Modified -= DotNetProject_Modified;
+
+            UpdateHostProjectForeground(null);
+        }
+
+        protected abstract Task OnProjectChangedAsync();
+
+        // Protected virtual for testing
+        protected virtual void AttachToProject()
+        {
+            ForegroundDispatcher.AssertForegroundThread();
+
+            DotNetProject.Modified += DotNetProject_Modified;
+
+            // Trigger the initial update to the project.
+            _batchingProjectChanges = true;
+            Task.Factory.StartNew(ProjectChangedBackgroundAsync, null, CancellationToken.None, TaskCreationOptions.None, ForegroundDispatcher.BackgroundScheduler);
+        }
+
+        // Must be called inside the lock.
+        protected async Task UpdateHostProjectUnsafeAsync(HostProject newHostProject)
+        {
+            if (_foregroundDispatcher.IsForegroundThread)
+            {
+                UpdateHostProjectForeground(newHostProject);
+            }
+            else
+            {
+                await Task.Factory.StartNew(UpdateHostProjectForeground, newHostProject, CancellationToken.None, TaskCreationOptions.None, ForegroundDispatcher.ForegroundScheduler);
+            }
```

Comment:
This seems a little more complicated than just requiring this to be foreground only

Reply:
Eh, `UpdateHostProjectUnsafeAsync` is always called on the background thread initially (we do time consuming work on the bg thread). I just added the if statement for the just-in-case situation. I could make it background thread only

Reply:
nm, will just require callers to use the scheduler for the foreground thread.